src/domains/transaction/hooks/use-pending-transactions.ts

```diff
@@ -2,103 +2,150 @@ import { useCallback } from "react";
+			}
+		},
+		[setPendingJson],
 	);
 
 	const addPendingTransaction = useCallback(
 		(transaction: DTO.ExtendedSignedTransactionData) => {
 			try {
-				const pendingTransaction: PendingTransactionData = {
-					convertedAmount: transaction.convertedAmount?.(),
-					convertedTotal: transaction.convertedTotal(),
+				const input = {
+					data: transaction.data().data().data,
 					explorerLink: transaction.explorerLink(),
-					fee: transaction.fee(),
 					from: transaction.from(),
 					hash: transaction.hash(),
-					isMultiPayment: transaction.isMultiPayment(),
-					isReturn: transaction.isReturn(),
-					isTransfer: transaction.isTransfer(),
-					isUnvote: transaction.isUnvote(),
-					isUpdateValidator: transaction.isUpdateValidator(),
-					isUsernameRegistration: transaction.isUsernameRegistration(),
-					isUsernameResignation: transaction.isUsernameResignation(),
-					isValidatorRegistration: transaction.isValidatorRegistration(),
-					isValidatorResignation: transaction.isValidatorResignation(),
-					isVote: transaction.isVote(),
-					isVoteCombination: transaction.isVoteCombination(),
 					networkId: transaction.wallet().networkId(),
 					nonce: transaction.nonce(),
-					recipients: transaction.recipients?.(),
-					timestamp: DateTime.make(transaction.timestamp()),
 					to: transaction.to(),
-					total: transaction.total(),
-					type: transaction.type(),
 					value: transaction.value(),
 					walletAddress: transaction.wallet().address(),
 				};
 
-				setPendingTransactions((prev) => {
-					const filtered = prev.filter((tx) => tx.hash !== transaction.hash());
-					return [...filtered, pendingTransaction];
-				});
+				addPendingTransactionFromUnconfirmed(input as any);
 			} catch (error) {
 				/* istanbul ignore next -- @preserve */
 				console.error("Failed to add pending transaction:", error);
 			}
 		},
-		[setPendingTransactions],
+		[addPendingTransactionFromUnconfirmed],
 	);
 
 	const removePendingTransaction = useCallback(
 		(hash: string) => {
-			setPendingTransactions((prev) => prev.filter((tx) => tx.hash !== hash));
+			setPendingJson((prev) => prev.filter((p) => p.hash !== hash));
+		},
+		[setPendingJson],
+	);
+
+	const buildPendingForUI = useCallback(
+		(walletAddresses: string[], wallets: Contracts.IReadWriteWallet[]) => {
+			const dtos = pendingJson.map(PendingDTO.fromPersistedJSON);
+
+			return dtos.map((dto) => {
+				const wallet = wallets.find((w) => w.address() === dto.getMeta("address"));
+				const isReceived = walletAddresses.includes(dto.to());
+				const isSent = walletAddresses.includes(dto.from());
+
+				return {
+					blockHash: () => {},
+					confirmations: () => BigNumber.make(0),
+					convertedAmount: () => 0,
+					convertedTotal: () => 0,
+					explorerLink: () => dto.getMeta<string>("explorerLink") ?? "",
+					fee: () => dto.fee(),
+					from: () => dto.from(),
+					hash: () => dto.hash(),
+					isConfirmed: () => false,
+					isFailed: () => false,
```

Comment:
how come the dto itself doesn't cover these methods? it feels off having to re-define them when majority is part of the dto already